File: training/csharp/src/De.Sots.CellarsAndCentaurs/Adapter/Persistence/PersistedCreatures.cs
using De.Sots.CellarsAndCentaurs.Application;
using De.Sots.CellarsAndCentaurs.Domain.Model;
using De.Sots.CellarsAndCentaurs.Domain.Service;

namespace De.Sots.CellarsAndCentaurs.Adapter.Persistence;

public class PersistedCreatures : Creatures
{
    private readonly CreatureRepository repository;

    public PersistedCreatures(CreatureRepository repository)
    {
        this.repository = repository;
    }

    public void Save(Creature creature)
    {
        repository.Add(new CreatureEntity(creature.Id.Value, creature.Type, creature.HitPoints?.Max ?? 0));
    }

    public Creature Find(CreatureId id)
    {
        var entity = repository.FindOne(id.Value) ?? throw new NoSuchCreatureException(id);
        return new Creature(new CreatureId(entity.Id), CreatureFacade.STANDARD_CREATURE_TYPE)
        {
            HitPoints = HitPoints.Init(entity.HitPoints)
        };
    }
}

File: sample-projects/vue/src/de/sots/cellarsandcentaurs/adapter/persistence/PersistedCreatures.ts
import { CreatureRepository } from './CreatureRepository'
import { CreatureEntity } from './CreatureEntity'
import { Creatures } from '../../domain/service/Creatures'
import { Creature } from '../../domain/model/Creature'
import { CreatureId } from '../../domain/model/CreatureId'
import { NoSuchCreatureException } from '../../domain/model/NoSuchCreatureException'
import { CreatureFacade } from '../../application'

export class PersistedCreatures implements Creatures {
  constructor(private readonly repository: CreatureRepository) {}

  async save(creature: Creature): Promise<void> {
    await this.repository.save(new CreatureEntity(creature.getId().id))
  }

  async find(id: CreatureId): Promise<Creature> {
    const entity = await this.repository.findOne(id.id)
    if (!entity) {
      throw new NoSuchCreatureException(id)
    }
    return new Creature(new CreatureId(entity.id), CreatureFacade.STANDARD_CREATURE_TYPE)
  }
}
